### ISSUE TITLE
Error display decimal divider for zero double value in Excel

### ISSUE THREAD

**Issue by QuarK264** [OfficeDev/Open-XML-SDK#527]
**Description**

I have an error connected with displaying decimal divider for zero double value in Excel using custom `NumberingFormat`. I set `FormatCode` `"0.####"`, it works fine for values like 3.4342, but fails for 0.0 (get 0.). I have tried to change `FormatCode` to `"0.0###"`, when I do it, I got sheet error when open document.

I think it is a bug, as if I change cell type to string and use pattern `"0.####"` in C# `ToString` method for 0.0 value is displayed 0.

**Information**

- .NET Target: .NET Core 2.0
- DocumentFormat.OpenXml Version: 2.7.1

**Repro**

```csharp
using (var document = SpreadsheetDocument.Create(filename, SpreadsheetDocumentType.Workbook))
{
    var workbookPart = document.AddWorkbookPart();
    workbookPart.Workbook = new Workbook();
    var worksheetPart = workbookPart.AddNewPart<WorksheetPart>();

    var sheetData = CreateContent();

    DefineStyles(document);
    worksheetPart.Worksheet = new Worksheet();
    worksheetPart.Worksheet.Append(sheetData);
    var sheets = document.WorkbookPart.Workbook.AppendChild(new Sheets());
    var sheet = new Sheet
    {
        Id = document.WorkbookPart.GetIdOfPart(worksheetPart),
        SheetId = 1,
        Name = Strings.SheetName
    };
    sheets.Append(sheet);

    workbookPart.Workbook.Save();
    document.Close();
}

private static SheetData CreateContent()
{
    var value = TimeSpan.Zero;
    var row1 = new Row {RowIndex = 1};
    var title = new Cell
    {
        CellReference = "A1",
        CellValue = new CellValue(value.TotalHours.ToString(CultureInfo.InvariantCulture)),
        DataType = CellValues.Number,
        StyleIndex = 1U
    };
    row1.Append(title);

    var sheetData = new SheetData();
   sheetData.Append(row1);

   return sheetData;
}

private static void DefineStyles(SpreadsheetDocument document)
{
    var stylesPart = document.WorkbookPart.AddNewPart<WorkbookStylesPart>();
    var stylesheet = new Stylesheet();
    
    var fonts = new Fonts();
    var font = new Font
    {
        FontName = new FontName
        {
            Val = "Calibri"
        },
        FontSize = new FontSize
        {
            Val = 11
        }
    };
    fonts.Append(font);
    fonts.Count = (uint) fonts.ChildElements.Count;

    var fills = new Fills();
    var fill = new Fill
    {
        PatternFill = new PatternFill
        {
            PatternType = PatternValues.None
        }
    };
    fills.Append(fill);
    fills.Count = (uint) fills.ChildElements.Count;

    var borders = new Borders();
    var border = new Border
    {
        LeftBorder = new LeftBorder(),
        RightBorder = new RightBorder(),
        TopBorder = new TopBorder(),
        BottomBorder = new BottomBorder(),
        DiagonalBorder = new DiagonalBorder()
    };
    borders.Append(border);
    borders.Count = (uint) borders.ChildElements.Count;

    var cellStyleFormats = new CellStyleFormats();
    var defaultCellFormat = new CellFormat
    {
        NumberFormatId = 0,
        FontId = 0,
        FillId = 0,
        BorderId = 0
    };
    cellStyleFormats.Append(defaultCellFormat);
    cellStyleFormats.Count = (uint) cellStyleFormats.ChildElements.Count;

    var numberingFormats = new NumberingFormats();
    //Style with bug
    var nFormat2Decimal = new NumberingFormat
    {
        NumberFormatId = 1U,
        FormatCode = StringValue.FromString("0.####")
    };
    numberingFormats.Append(nFormat2Decimal);

    var cellFormats = new CellFormats();
    defaultCellFormat = new CellFormat
    {
        NumberFormatId = 0,
        FontId = 0,
        FillId = 0,
        BorderId = 0,
        FormatId = 0
    };
    cellFormats.Append(defaultCellFormat);

    var timeDoubleCellFormat = new CellFormat
    {
        FontId = 0,
        FillId = 0,
        BorderId = 0,
        FormatId = 1U,
        NumberFormatId = 1U,
        ApplyNumberFormat = true
    };
    cellFormats.Append(timeDoubleCellFormat);

    var cellStyles = new CellStyles();
    var cellStyle = new CellStyle
    {
        Name = "Normal",
        FormatId = 0,
        BuiltinId = 0
    };
    cellStyles.Append(cellStyle);
    cellStyles.Count = (uint) cellStyles.ChildElements.Count;

    var differentialFormats = new DifferentialFormats {Count = 0};

    var tableStyles = new TableStyles
    {
        Count = 0,
        DefaultTableStyle = "TableStyleMedium9",
        DefaultPivotStyle = "PivotStyleLight16"
    };

    stylesheet.Append(numberingFormats);
    stylesheet.Append(fonts);
    stylesheet.Append(fills);
    stylesheet.Append(borders);
    stylesheet.Append(cellStyleFormats);
    stylesheet.Append(cellFormats);
    stylesheet.Append(cellStyles);
    stylesheet.Append(differentialFormats);
    stylesheet.Append(tableStyles);

    stylesPart.Stylesheet = stylesheet;
    stylesPart.Stylesheet.Save();
}
``` 
Sorry for long style definition, it seems it all is mandatory.

**Observed**

![CurrentResult](https://user-images.githubusercontent.com/9616488/48188153-1d1a0980-e34e-11e8-97f4-7947a9b31a4c.jpg)

**Expected**

![ExpectedResult](https://user-images.githubusercontent.com/9616488/48188500-f3151700-e34e-11e8-874b-ab63b508de1e.PNG)

**Comment by tomjebo** [comment]
@QuarK264, thanks for reporting this issue. I’ll repro and investigate.

**Comment by QuarK264** [comment]
@tomjebo I don't understand, what do you mean, when you talk about ';' separated sections. I don't use them or I don't know about them.

Format code `"#,##0.0000"` doesn't fit to my situation, as I don't need thousand separator. I need the next format: at least zero or more figures before decimal separator (dot) and at least zero or up to 4 figures after dot. I see two possible formats for my case `"0.####"` and `"0.0###"`. For floating point value 0.0 applying first format gives me string `"0."` (zero after dot is missing), applying second format gives me error when I am trying to open excel document, after its restore my numbers are restored as strings. First format may look strange, but I think it is possible, as native method `string.Format` works with it fine.

I hope, I described my problem more clearly.

**Comment by tomjebo** [comment]
@QuarK264 if you check the standard section I pointed to you'll see the ";" thing I mentioned. But anyway, I tested the following: 

`  <x:numFmts>
    <x:numFmt numFmtId="1" formatCode="0.0###"/>
  </x:numFmts>`

And it works just fine for a value of 0.0. 

![image](https://user-images.githubusercontent.com/4644198/50652703-2f728700-0f3c-11e9-9eeb-64150effa665.png)

And using "0.####" will indeed give you "0." in the cell because (check the definition of numFmt formatCode attribute) "#" means that extra 0's won't be displayed if there are fewer digits than the number of #'s. 

If you are getting an error opening the workbook, it could be due to the version of Excel you're using. I'm using the latest but haven't tried previous versions. 
In any case, the format string is correct but if Excel is behaving badly, this should be reported on the Excel usage support forums. It could be a bug in some version of Excel.
